File: src/shipinfer/core/events/schema.py
"""The perception event: one frame's finished answer, as the wire and the file see it.

Schema v4: per-object parallel arrays (``*_bbox_vec``, ``*_track_id_vec``,
``*_global_id_vec``, ``*_feature_vec``) split by class, plus frame identity (``camera_id``,
``image_id``, ``sub_id``), geometry (``img_width/height/fps``) and ``missing_stages`` — a partial frame
says so instead of reading as an empty complete one. Why every v1 key keeps its name, type
and people-only meaning (a deployed ``motservice`` must need no rebuild):
``docs/design/event-schema.md``. Stdlib only, by construction and by test
(``TestTheSchemaIsPortable``): a consumer may copy this module out wholesale.
``PerceptionEvent.build`` is the one constructor; ``pipeline/schema.py`` re-exports for the
old spelling.
"""

from __future__ import annotations

import json
import time
from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import Any

__all__ = [
    "MESSAGE_TYPE",
    "SCHEMA_VERSION",
    "ObjectRecord",
    "PerceptionEvent",
]

#: The ``type`` field every consumer switches on. Unchanged from v1 on purpose: the value is
#: part of the contract, and a new one would be routed nowhere by a deployed consumer.
MESSAGE_TYPE = "Det2MOT"

#: 1 was ``DetectionMOTFrameData``. 2 adds ships, timing and completeness; 3 adds the
#: track id and its state; 4 adds the cross-camera ``global_id``. Every step is additive, and
#: the number is bumped rather than left alone precisely so a consumer can branch on it
#: instead of probing for a key.
SCHEMA_VERSION = 4


@dataclass(frozen=True, slots=True)
class ObjectRecord:
    """One detected object and everything the DAG learned about it.

    ``frozen`` and ``slots`` because there are 15 000 of these a second: this is the
    highest-cardinality value in the system, and a per-object dict would cost more than the
    inference that produced it.

    A field is ``None`` when the stage that fills it did not run — a person has no
    ``ship_id`` and never will, and a ship whose recogniser timed out has none *yet*. Both
    are distinguishable from ``0``, which is a legitimate gallery id.
    """

    #: ``<camera>_<frame>_<index>``: unique across the fleet and derivable by a consumer, so
    #: a tracker can key on it without a side channel.
    det_id: str
    #: The label from ``pipeline.class_labels`` — ``"ship"``, ``"person"``, or whatever a
    #: deployment configured. Never a raw class index: the index is a property of the
    #: checkpoint and changes when the model is retrained.
    class_name: str
    score: float
    #: ``(x1, y1, x2, y2)`` in the **source frame's** pixels, letterbox already undone.
    bbox: tuple[float, float, float, float]
    #: Appearance embedding, empty when the embedder did not run for this object.
    embedding: tuple[float, ...] = ()
    #: Gallery identity, ships only.
    ship_id: int | None = None
    #: Cosine similarity behind :attr:`ship_id`, so a consumer can apply its own threshold
    #: instead of trusting ours.
    similarity: float | None = None
    #: Mask area in pixels. The mask *itself* is deliberately not published: a 512x512
    #: float mask is 1 MB and this bus carries metadata, not pixels — the architecture doc
    #: is explicit that frames stay in shared memory and Kafka gets the small results.
    mask_area_px: float | None = None
    #: Single-camera track identity, from Plane 3 — process-unique so two cameras' tracklets
    #: meet in the cross-camera tier without colliding. ``None`` = tracking off, the track not
    #: yet confirmed (an identity that dies after two frames never existed downstream), or the
    #: frame lost the ordering race; distinct from id 0, which the counter never issues.
    track_id: int | None = None
    #: That track's lifecycle state, as the tracker reported it. Carried rather than assumed
    #: because a consumer that has to trust our filtering cannot apply its own.
    track_state: str | None = None
    #: Fleet-wide identity for this object's tracklet, from the cross-camera tier (v4). The
    #: same object seen by two cameras carries the same ``global_id`` and two different
    #: ``track_id``\ s, which is the whole distinction between the two fields: one is unique
    #: within a camera's timeline, the other across the deployment.
    #:
    #: ``None`` when cross-camera association is off, when the frame's instant closed without
    #: this camera (``mtmc`` in :attr:`PerceptionEvent.missing_stages`), or when this object
    #: had no published track for an id to attach to. Distinguishable from ``0``, which the
    #: assigner never issues.
    global_id: int | None = None

    @property
    def bbox_list(self) -> list[float]:
        """The box in the wire format's shape (a JSON array, as in v1)."""
        return [float(v) for v in self.bbox]


@dataclass(frozen=True, slots=True)
class PerceptionEvent:
    """One frame's perception result, ready to serialise.

    Immutable because it is handed to a sink that may batch, retry or fan it out, and a
    value that can be mutated after being queued for publication is a race with no symptom
    beyond a wrong number in a dashboard.
    """

    camera_id: str
    #: The frame id from the ingest tag. Named ``image_id`` on the wire, as in v1.
    frame_id: int
    #: Which perception process produced this — v1's ``sub_id``.
    source_id: str
    objects: tuple[ObjectRecord, ...] = ()
    img_width: int = 0
    img_height: int = 0
    img_fps: int = 0
    #: Wall-clock capture time, for a human and for a consumer joining across processes.
    captured_unix_ns: int = 0
    emitted_unix_ns: int = 0
    #: Capture to emission. Measured from the monotonic clock, which is why it is carried
    #: rather than derived from the two wall-clock stamps: NTP stepping the clock must not
    #: turn into a negative latency in a dashboard.
    latency_us: int = 0
    #: Stages that never delivered. Empty for a complete frame.
    missing_stages: tuple[str, ...] = ()
    #: Why this event was emitted, in the collector's own five words: ``complete``,
    #: ``incomplete``, ``timeout``, ``shutdown``, ``evicted``, passed through verbatim by
    #: ``pipeline/runner.py``. It used to say ``failed``, which nothing emits -- and the C++
    #: port read this line and wrote it for two of the five (fixed with P5-A).
    reason: str = "complete"
    schema_version: int = SCHEMA_VERSION
    type: str = MESSAGE_TYPE
    #: Free-form additions a deployment needs and the schema should not grow a field for.
    extra: dict[str, Any] = field(default_factory=dict)

    # -- construction ------------------------------------------------------------------

    @classmethod
    def build(
        cls,
        *,
        camera_id: str,
        frame_id: int,
        source_id: str,
        objects: Sequence[ObjectRecord],
        width: int = 0,
        height: int = 0,
        fps: float = 0.0,
        captured_ns: int = 0,
        captured_unix_ns: int = 0,
        missing_stages: Sequence[str] = (),
        reason: str = "complete",
    ) -> PerceptionEvent:
        """Stamp an event with both clocks read at the moment of emission.

        ``captured_ns`` is monotonic and ``captured_unix_ns`` is wall clock, exactly as the
        ingest tag carries them, and the latency is computed from the monotonic pair only.
        """
        now_ns = time.monotonic_ns()
        return cls(
            camera_id=camera_id,
            frame_id=frame_id,
            source_id=source_id,
            objects=tuple(objects),
            img_width=int(width),
            img_height=int(height),
            img_fps=round(fps),
            captured_unix_ns=captured_unix_ns,
            emitted_unix_ns=time.time_ns(),
            latency_us=max(0, (now_ns - captured_ns) // 1000) if captured_ns else 0,
            missing_stages=tuple(missing_stages),
            reason=reason,
        )

    # -- views -------------------------------------------------------------------------

    @property
    def key(self) -> tuple[str, int]:
        """``(camera_id, frame_id)`` — the tag this event must carry unchanged (ADR-002)."""
        return (self.camera_id, self.frame_id)

    @property
    def is_partial(self) -> bool:
        return bool(self.missing_stages)

    def objects_of(self, class_name: str) -> tuple[ObjectRecord, ...]:
        """Every object of one class, in detection order."""
        return tuple(o for o in self.objects if o.class_name == class_name)

    def as_det2mot(self) -> dict[str, Any]:
        """Exactly the v1 ``Det2MOT`` payload: people only, v1 key names, v1 types.

        Byte-compatible with ``DetectionMOTFrameData::serialize`` so a deployed
        ``motservice`` consumes it unchanged. Nothing new appears here — a v1 consumer that
        validates its input strictly must see a v1 message.
        """
        people = self.objects_of("person")
        return {
            "sub_id": self.source_id,
            "det_id_vec": [o.det_id for o in people],
            "camera_id": self.camera_id,
            "image_id": self.frame_id,
            "det_body_score_vec": [o.score for o in people],
            "body_bbox_vec": [o.bbox_list for o in people],
            "body_feature_vec": [o.embedding for o in people],
            "img_width": self.img_width,
            "img_height": self.img_height,
            "img_fps": self.img_fps,
        }

    def as_dict(self) -> dict[str, Any]:
        """The current payload: every v1 key, plus ships, timing, completeness and tracklets."""
        ships = self.objects_of("ship")
        people = self.objects_of("person")
        payload = self.as_det2mot()
        payload.update(
            {
                "type": self.type,
                "schema_version": self.schema_version,
                # Ships in the same parallel-array idiom as the person fields above, so a
                # consumer written against v1 reads v2 with the same helpers.
                "ship_det_id_vec": [o.det_id for o in ships],
                "det_ship_score_vec": [o.score for o in ships],
                "ship_bbox_vec": [o.bbox_list for o in ships],
                "ship_feature_vec": [o.embedding for o in ships],
                "ship_id_vec": [o.ship_id for o in ships],
                "ship_similarity_vec": [o.similarity for o in ships],
                "ship_mask_area_vec": [o.mask_area_px for o in ships],
                # Tracklets (v3). People keep the unprefixed ``body_`` idiom the v1 person
                # arrays use and ships keep the ``ship_`` one, so a consumer that already
                # walks one set of parallel arrays walks these with the same helper. A null
                # entry means this object has no published identity — see ObjectRecord.
                "body_track_id_vec": [o.track_id for o in people],
                "body_track_state_vec": [o.track_state for o in people],
                "ship_track_id_vec": [o.track_id for o in ships],
                "ship_track_state_vec": [o.track_state for o in ships],
                # Cross-camera identity (v4), one entry per object, beside the per-camera
                # track id rather than replacing it: a consumer that joins two cameras reads
                # these, and one that only follows a single camera keeps reading the pair
                # above. A null entry means this object has no global identity — see
                # ObjectRecord.global_id for the three ways that happens.
                "body_global_id_vec": [o.global_id for o in people],
                "ship_global_id_vec": [o.global_id for o in ships],
                "captured_unix_ns": self.captured_unix_ns,
                "emitted_unix_ns": self.emitted_unix_ns,
                "latency_us": self.latency_us,
                # A frame with no people and a frame whose embedder timed out look
                # identical in v1. These two keys are the difference.
                "partial": self.is_partial,
                "missing_stages": list(self.missing_stages),
                "reason": self.reason,
            }
        )
        if self.extra:
            payload["extra"] = dict(self.extra)
        return payload

    def to_json(self) -> str:
        """One line of JSON. ``separators`` because 1000 of these a second is 1000 x the
        whitespace, and every byte crosses a broker."""
        return json.dumps(self.as_dict(), separators=(",", ":"))

    def __repr__(self) -> str:
        state = f" partial missing={list(self.missing_stages)}" if self.is_partial else ""
        return (
            f"<PerceptionEvent cam={self.camera_id} frame={self.frame_id} "
            f"objects={len(self.objects)}{state}>"
        )

File: csrc/shipinfer/core/events/schema.cpp
#include "shipinfer/core/events/schema.h"

#include <charconv>
#include <chrono>
#include <cmath>

#include "shipinfer/core/events/json.h"

namespace shipinfer::events {

    namespace {

        // A JSON array, one element per object of one class. Every `*_vec` key in the payload
        // is this shape, which is why it is a template rather than eleven near-identical loops.
        //
        // Appends, like everything else here: a returned `std::string` per element meant one
        // allocation per number, and an embedding is 2048 of them (P5-A-ALLOC, measured by
        // `cli/bench_events.cpp`).
        template <typename Fn>
        void append_array(std::string& out, const std::vector<const ObjectRecord*>& objects,
                          Fn&& element) {
            out += '[';
            for (size_t i = 0; i < objects.size(); ++i) {
                if (i) out += ',';
                element(out, *objects[i]);
            }
            out += ']';
        }

        void append_optional_int(std::string& out, const std::optional<int64_t>& value) {
            // `null`, not a sentinel: the Python field is `int | None` and 0 is a legitimate
            // gallery id, a legitimate track id and a legitimate global id.
            if (!value) {
                out += "null";
                return;
            }
            char buffer[24];
            const std::to_chars_result done =
                std::to_chars(buffer, buffer + sizeof(buffer), *value);
            out.append(buffer, done.ptr);
        }

        void append_int(std::string& out, int64_t value) {
            char buffer[24];
            const std::to_chars_result done =
                std::to_chars(buffer, buffer + sizeof(buffer), value);
            out.append(buffer, done.ptr);
        }

        void append_optional_double(std::string& out, const std::optional<double>& value) {
            if (value) {
                append_number(out, *value);
            } else {
                out += "null";
            }
        }

        void append_optional_word(std::string& out, const std::optional<std::string>& value) {
            if (value) {
                append_string(out, *value);
            } else {
                out += "null";
            }
        }

        void append_bbox(std::string& out, const ObjectRecord& object) {
            out += '[';
            for (int i = 0; i < 4; ++i) {
                if (i) out += ',';
                append_number(out, object.bbox[i]);
            }
            out += ']';
        }

        void append_embedding(std::string& out, const ObjectRecord& object) {
            out += '[';
            for (size_t i = 0; i < object.embedding.size(); ++i) {
                if (i) out += ',';
                append_number(out, object.embedding[i]);
            }
            out += ']';
        }

        void append_words(std::string& out, const std::vector<std::string>& values) {
            out += '[';
            for (size_t i = 0; i < values.size(); ++i) {
                if (i) out += ',';
                append_string(out, values[i]);
            }
            out += ']';
        }

        // What to ask for up front. Deliberately generous rather than exact: one over-large
        // `reserve` costs a page or two of untouched address space, while an under-estimate
        // costs a reallocation and a 400 KB copy on the emission path. 14 bytes is the
        // measured mean for a `repr`-spelled double plus its comma.
        size_t estimated_bytes(const std::vector<ObjectRecord>& objects) {
            size_t total = 1024;
            for (const ObjectRecord& object : objects) {
                total += 320 + 14 * (object.embedding.size() + 4);
            }
            return total;
        }

    }  // namespace

    std::vector<const ObjectRecord*> PerceptionEvent::objects_of(
        const std::string& class_name) const {
        std::vector<const ObjectRecord*> found;
        for (const ObjectRecord& object : objects) {
            if (object.class_name == class_name) found.push_back(&object);
        }
        return found;
    }

    // doc: long the key ORDER is the contract, so it is written out rather than assembled
    std::string PerceptionEvent::to_json() const {
        // In `as_det2mot()`'s order first and then `as_dict()`'s additions, because a Python
        // dict preserves insertion order and `update()` appends keys it does not already
        // hold -- so this sequence IS the byte sequence the other plane produces. Assembling
        // it from a map would sort the keys and the gate would fail on every event.
        const std::vector<const ObjectRecord*> people = objects_of("person");
        const std::vector<const ObjectRecord*> ships = objects_of("ship");
        const auto det_id = [](std::string& to, const ObjectRecord& o) {
            append_string(to, o.det_id);
        };
        const auto score = [](std::string& to, const ObjectRecord& o) {
            append_number(to, o.score);
        };
        const auto ship_id = [](std::string& to, const ObjectRecord& o) {
            append_optional_int(to, o.ship_id);
        };
        const auto similarity = [](std::string& to, const ObjectRecord& o) {
            append_optional_double(to, o.similarity);
        };
        const auto mask_area = [](std::string& to, const ObjectRecord& o) {
            append_optional_double(to, o.mask_area_px);
        };
        const auto track_id = [](std::string& to, const ObjectRecord& o) {
            append_optional_int(to, o.track_id);
        };
        const auto track_state = [](std::string& to, const ObjectRecord& o) {
            append_optional_word(to, o.track_state);
        };
        const auto global_id = [](std::string& to, const ObjectRecord& o) {
            append_optional_int(to, o.global_id);
        };

        // ONE buffer for the whole event, reserved once. Every helper above appends into it,
        // so a 2048-float embedding costs 2048 `to_chars` calls and no allocation at all --
        // where it used to cost about three per number.
        std::string out;
        out.reserve(estimated_bytes(objects));
        out += '{';
        out += "\"sub_id\":";
        append_string(out, source_id);
        out += ",\"det_id_vec\":";
        append_array(out, people, det_id);
        out += ",\"camera_id\":";
        append_string(out, camera_id);
        out += ",\"image_id\":";
        append_int(out, frame_id);
        out += ",\"det_body_score_vec\":";
        append_array(out, people, score);
        out += ",\"body_bbox_vec\":";
        append_array(out, people, append_bbox);
        out += ",\"body_feature_vec\":";
        append_array(out, people, append_embedding);
        out += ",\"img_width\":";
        append_int(out, img_width);
        out += ",\"img_height\":";
        append_int(out, img_height);
        out += ",\"img_fps\":";
        append_int(out, img_fps);
        out += ",\"type\":";
        append_string(out, kMessageType);
        out += ",\"schema_version\":";
        append_int(out, kSchemaVersion);
        out += ",\"ship_det_id_vec\":";
        append_array(out, ships, det_id);
        out += ",\"det_ship_score_vec\":";
        append_array(out, ships, score);
        out += ",\"ship_bbox_vec\":";
        append_array(out, ships, append_bbox);
        out += ",\"ship_feature_vec\":";
        append_array(out, ships, append_embedding);
        out += ",\"ship_id_vec\":";
        append_array(out, ships, ship_id);
        out += ",\"ship_similarity_vec\":";
        append_array(out, ships, similarity);
        out += ",\"ship_mask_area_vec\":";
        append_array(out, ships, mask_area);
        out += ",\"body_track_id_vec\":";
        append_array(out, people, track_id);
        out += ",\"body_track_state_vec\":";
        append_array(out, people, track_state);
        out += ",\"ship_track_id_vec\":";
        append_array(out, ships, track_id);
        out += ",\"ship_track_state_vec\":";
        append_array(out, ships, track_state);
        out += ",\"body_global_id_vec\":";
        append_array(out, people, global_id);
        out += ",\"ship_global_id_vec\":";
        append_array(out, ships, global_id);
        out += ",\"captured_unix_ns\":";
        append_int(out, captured_unix_ns);
        out += ",\"emitted_unix_ns\":";
        append_int(out, emitted_unix_ns);
        out += ",\"latency_us\":";
        append_int(out, latency_us);
        out += ",\"partial\":";
        out += is_partial() ? "true" : "false";
        out += ",\"missing_stages\":";
        append_words(out, missing_stages);
        out += ",\"reason\":";
        append_string(out, reason);
        if (!extra.empty()) {
            out += ",\"extra\":{";
            bool first = true;
            for (const auto& [key, value] : extra) {
                if (!first) out += ',';
                first = false;
                append_string(out, key);
                out += ':';
                append_string(out, value);
            }
            out += '}';
        }
        out += '}';
        return out;
    }

    PerceptionEvent build(const std::string& camera_id, int64_t frame_id,
                          const std::string& source_id, std::vector<ObjectRecord> objects,
                          int64_t width, int64_t height, double fps, int64_t captured_ns,
                          int64_t captured_unix_ns,
                          const std::vector<std::string>& missing_stages,
                          const std::string& reason, int64_t now_ns, int64_t now_unix_ns) {
        using namespace std::chrono;
        PerceptionEvent event;
        event.camera_id = camera_id;
        event.frame_id = frame_id;
        event.source_id = source_id;
        event.objects = std::move(objects);
        event.img_width = width;
        event.img_height = height;
        // `llrint` and NOT `llround`, matching `img_fps=round(fps)`: `llround` is half AWAY
        // FROM ZERO while Python's `round` is half TO EVEN, so they part company at exactly
        // the rates a deployment configures -- 12.5 is 13 there and 12 here. A fleet split
        // across Python and C++ shards would publish two `img_fps` for one camera, in the
        // one field this seam added a byte gate to protect. `llrint` uses the current
        // rounding mode, which is `FE_TONEAREST` (half to even) and never changed here.
        event.img_fps = static_cast<int64_t>(std::llrint(fps));
        event.captured_unix_ns = captured_unix_ns;
        // Injectable, and defaulted rather than required, because the parity gate has to
        // produce the same bytes twice and a wall clock cannot. Production passes nothing.
        event.emitted_unix_ns =
            now_unix_ns
                ? now_unix_ns
                : duration_cast<nanoseconds>(system_clock::now().time_since_epoch()).count();
        const int64_t monotonic =
            now_ns ? now_ns
                   : duration_cast<nanoseconds>(steady_clock::now().time_since_epoch()).count();
        event.latency_us =
            captured_ns ? std::max<int64_t>(0, (monotonic - captured_ns) / 1000) : 0;
        event.missing_stages = missing_stages;
        event.reason = reason;
        return event;
    }

}  // namespace shipinfer::events
